feat: import of @angular/service-worker

alxhub opened the pull request:
Imports the existing repository (without history) from
https://github.com/angular/mobile-toolkit. Sets up the
package to build using build.sh and to run unit tests
as part of test.sh. An integration test is also added,
wrapping the SW e2e test so that `npm run test` works.

chuckjaz committed:
fix(compiler): do not report type errors for arguments with `@Inject` (#16222)

Fixes #15424

chuckjaz committed:
fix(tsc-wrapped): ignore `|null` and `|undefined` when collecting types (#16222)

wardbell committed:
docs(aio): i18n guide - updates for v4

trotyl committed:
test(compiler): fix typo "mamespace"

tbosch committed:
refactor(compiler): make `OutputAst` contain the moduleName, not the filePath (#16832).

The goal of this change is to simplify the emitters,
as we will soon create a new one to emit TypeScript nodes directly.

heathkit committed:
fix(upgrade): call setInterval outside the Angular zone

This wraps the $interval service when using upgrade to run the
$interval() call outside the Angular zone. However, the callback is
invoked within the Angular zone, so changes still propagate to
downgraded components.

tbosch committed:
refactor(compiler): emit OutputAst and not sources (#16832)

This is in preparation for creating typescript nodes
directly from `OutputAst` nodes.

tbosch committed:
refactor(compiler): allow sync AOT compilation (#16832).

AOT compilation can be executed synchronously now,
if the `ReosurceLoader` returns a string directly
(and no `Promise`).

tbosch committed:
ci: add vicb as owner to compiler-cli (#16832)

tbosch committed:
refactor(compiler): don’t rely on global reflector (#16832)

Using the global reflector made it impossible
to compile multiple programs at the same time.

wardbell committed:
feat(aio): add progress bar

closes #16110

gkalpak committed:
test(aio): avoid warnings from Material during tests

juliemr committed:
feat(core): update zone.js to 0.8.10 and expose the flush method (#16860)

`flush()` can now be used from within fakeAsync tests to simulate moving
time forward until all macrotask events have been cleared from the
event queue.

ocombe committed:
feat(compiler): add location note to extracted xliff2 files (#16791)

Add source location as a note tag as `<note category="location">path/to/file.ts:start_line[,end_line]</note>`.
`[,end_line]` part is optional and specified only if the end line is different from the start line.

Fixes  #16531

chuckjaz committed:
docs(aio): document the non-null assertion operator

mhevery committed:
refactor(platform-browser): Remove setGlobalVar from DOM adapter

matsko committed:
fix(animations): repair flicker issues with WA polyfill (#16937)

Fixes #16919
Fixes #16918

vicb committed:
docs: update the release schedule (#16942)

vicb committed:
docs: update the release schedule (#16943)

IgorMinar committed:
build(aio): start using tsc importHelpers option for the app

this means we'll be temporarily duplicating the helpers (onces included via scripts
and secondly imported via es imports) - once rxjs, core and material migrate over
to tslib, we can drop the scripts/global dupe.

IgorMinar committed:
build(aio): upgrade ngo to 0.0.3 with static prop unhoisting

gkalpak committed:
ci(aio): deploy commits on the stable branch to production

The current stable branch is determined based on the current version mapped to
the npm `latest` tag (by replacing the patch version number with 'x' - e.g.
`1.2.3` --> `1.2.x`).
PRs against the stable branch will be deployed to the preview server (as long as
the rest of the requirements are met). Commits on the stable branch itself will
be deployed to production.

Fixes #16908

petebacondarwin committed:
fix(aio): use a special version of the TopBar if the screen is narrow

If there is a `TopBarNarrow` nav view then use this when the screen is narrow.
Otherwise just use the normal `TopBar`.

This commit also creates such a narrow topbar view where the "Docs" item is
in a different position

Closes #16940

tbosch committed:
fix(core): make decorators closure safe (#16905)

This is required as e.g. `token` from `@Inject` is
accessed in string form via makeParamDecorator
but as a property in the `ReflectiveInjector`.

Closes #16889 as this is a more general fix.

jasonaden committed:
build: import tslib rather than output TS helpers (#16901)

tbosch committed:
refactor(compiler): add ability to produce stub .ngfactory / .ngsummary files (#16963)

These files are needed so that:
- user code can compile even without real codegen
- as tsc transformers cannot create but only change existing files
  in the transformation pipeline.

gkalpak committed:
fix(aio): align top-level nav-items consistently

Fixes #16939

petebacondarwin committed:
build(aio): better error message for invalid links (#16993)

petebacondarwin committed:
aio: add h1 title to floating table of contents (#16959)

* refactor(aio): use explicit CSS class for TOC container

This makes the styling less fragile to changes in the HTML

* fix(aio): schedule TocComponent.activeIndex updates via AsapScheduler

We use the `asap` scheduler because updates to `activeItemIndex` are triggered by DOM changes,
which, in turn, are caused by the rendering that happened due to a ChangeDetection.

Without asap, we would be updating the model while still in a ChangeDetection handler,
which is disallowed by Angular.

* refactor(aio): do not instantiate floating ToC if not displayed

* feat(aio): display the h1 at the top of the floating TOC

Closes #16900

* refactor(aio): combine the TOC booleans flags into a "type" state

* refactor(aio): remove unnecessary `hostElement` property

* fix(aio): ensure that transition works on TOC

* fix(aio): use strict equality in ToC template

IgorMinar committed:
build(aio): upgrade purify to 0.0.11 with better coverage

IgorMinar committed:
build(aio): turn on pure_getters option in uglify

IgorMinar committed:
build(aio): use purify to include tslib

cli's scripts setup evals strings which requires too much escaping that bloats the code

sjtrimble committed:
feat(aio): marketing design refresh

- Add styling for active TOC item
- Add sidenav styles
- Change header tags to divs from index marketing page to remove anchors
- Fix use of card mixin and create separate card-docs class
- Add marketing styling
- Topnav styling when on home landing page
- Remove hamburger menu on home page
- Add fully rounded border to top nav toolbar search input
- Add mobile styles
- Add title banner to other marketing pages

tbosch committed:
fix(compiler): compile `.ngfactory.ts` files even if nobody references them. (#16899)

This is especially important for library authors, as they will
not reference the .ngfactory.ts files.

Fixes #16741

gkalpak committed:
fix(aio): correctly handle re-navigation to the empty path (#16997)

Foxandxss committed:
build(aio): less verbose boilerplate generator (#16970)

Foxandxss committed:
docs(aio): restore missing files (#16995)

petebacondarwin committed:
fix(aio): group topbar items into a single sidenav node when narrow screen (#17001)

The `TopBarNarrow` now only shows a single top level container, "About Angular",
and the original `TopBar` items will be children of this container.

The `TopBarNarrow` styling is changed to match the rest of the `SideNav`.

rspier committed:
fix(router): opening links in new window

Shift-clicks on router-links should not prevent browser default action.

A follow on to:
https://github.com/angular/angular/commit/1ac9dda93d01e5e61a3af32a9c596e6f26d50b00

matsko committed:
fix(animations): use a lightweight renderer for non-animation components (#17003)

tbosch committed:
refactor(compiler): change `bundle` to `flatModuleIndex` in the code (#17028)

tbosch committed:
refactor(compiler): change `bundle` to `flatModuleIndex` in the code (#17028)

petebacondarwin committed:
feat(aio): provide noop service worker file

Use this file if we need to turn off the service worker in deployed sites
in an emergency where the worker has a bug that is blocking the application
from working.

Closes #16897

IgorMinar committed:
ci(aio): switch over to shared account for deployment to staging and production

IgorMinar committed:
build(aio): externalize @angular/cli patches into a js file

This improves code readability and maintainability.

IgorMinar committed:
build(aio): add google webmaster tools verification file

this file needs to remain in the repo for us to remain verified as owners of aio-staging.firebaseapp.com.

petebacondarwin committed:
build(aio): upgrade to dgeni-packages 0.19.1

This provides the fix where static class members were not being displayed
in API docs, closes #17024

petebacondarwin committed:
build(aio): display "call" and "new" members of interfaces in API docs

chuckjaz committed:
Revert "refactor(compiler): change `bundle` to `flatModuleIndex` in the code (#17028)"

This reverts commit 5aa53d70aa8ee95a23adb3c7103346a6e68a5c2a.

chuckjaz committed:
Revert "fix(animations): use a lightweight renderer for non-animation components (#17003)"

This reverts commit 3ab86bd66162ca94de46e90d2754fcdf96f78c51.

jasonaden committed:
fix(router): make remove trailing slash consistent with URL params

closes #16069

IgorMinar committed:
feat(aio): temporarily add debugging code for google crawler (#17046)

Foxandxss committed:
docs(aio): update headers for toc (#16969)

matsko committed:
fix(animations): use a lightweight renderer for non-animation components (#17003)

This reverts commit c0981b8e13556e0481332d11f9739890593cd447.

chuckjaz committed:
release: cut the 4.2.0-rc.1 release

chuckjaz committed:
docs: add changelog for 4.2.0-rc.1

IgorMinar committed:
fix(aio): remove es2015 syntax from the onerror webmaster tools debugger (#17056)

IgorMinar committed:
fix(aio): add polyfill for GoogleBot (#17062)

I chose to use unknown=polyfill option as per recommendation found at https://github.com/Financial-Times/polyfill-service/issues/852#issuecomment-272097936

IgorMinar committed:
build(aio): upgrade to ngo with top-level NewExpression and CallExpression prefixing (#17063)

IgorMinar committed:
fix(aio): googlebot polyfill fix

gkalpak committed:
docs(aio): fix broken example in upgrade guide

Fixes #17109

jasonaden committed:
test(platform-server): fix target of TS outDir for integration test (#17118)

chuckjaz committed:
feat(compiler): emit typescript nodes from an output ast (#16823)

ocombe committed:
fix(common): always use 'other' case for locales with no plural rules (#16990)

Locales with no rules were using the wrong plural rule instead of the default.

devversion committed:
feat(tsc-wrapped): always convert shorthand imports (#16898)

Now converts shorthand imports for every TypeScript target. Tsickle is able to expand index shorthand imports for every TypeScript target and module possibility.

Expanding shorthand imports for CommonJS modules is also helpful when testing in the browser. Module loaders like SystemJS are not able to understand directory imports (or index shorthand imports)

petebacondarwin committed:
build(aio): capture the h1 title and attach it to the document

The HTML post-processing now collects any h1
that is found in the renderedContent and attaches
it to the doc via the `doc.vFile.title` property.

petebacondarwin committed:
build(aio): use the captured h1 as the title if necessary

If the doc does not already have a `title` tag
then use the one captured from the renderedContent in
the final JSON output, instead.

petebacondarwin committed:
build(aio): the the captured h1 as the title for the search index

If there is no title already provided, use the one captured from the renderedContent.

hotforfeature committed:
fix(compiler): enableLegacyTemplate should not be ignored (#17051)

Fixes #15555

chuckjaz committed:
build: update router package.json placeholder (#17060)

gkalpak committed:
fix(aio): fix ToC styling (#17110)

There seems to have been a bad rebase of #16228 on top of #16959, which affected
ToC styles from both PRs. This commit restores the horizontal line under `.h1`
elements and the vertical blue bar on the left-hand side of the ToC (with the
circle running along the bar to indicate the active section).

Fixes #17098

gkalpak committed:
fix(aio): fix SideNav items alignment (#17106)

Fixes #17097

gkalpak committed:
fix(aio): make the search-pane larger (and improve the search-box) (#17105)

* fix(aio): make the search-pane larger

Fixes #17094

* feat(aio): give the search-box a type of "search"

This enables browsers to style it better (e.g. add an `x` button for clearing
the field, which allows users to quickly reset the search query and hide
results).

IgorMinar committed:
fix(aio): disable search on browsers that don't support web workers

IgorMinar committed:
fix(aio): fix typo in web worker check (#17133)

gkalpak committed:
fix(aio): fix ToC styling

- Make the left bar gray (instead of blue).
- Show gray dot when hovering over an element.
- Hide left bar and dots in embedded mode.

gkalpak committed:
fix(aio): place progress bar at the top

Previously, the progress bar would be placed right under the static top bar. Now
that the top bar i not tatic any more, it makes more sense to place the progress
bar at the top of the page.

Fixes #17103

petebacondarwin committed:
build(aio): fix up API doc-gen templates

* Remove whitespace before type specifiers
* Generate `new` and `call` member info for interfaces
* Ensure that there is no double space after class names

vicb committed:
docs: add details about the WeakMap polyfill for 4.x in the changelog (#17146)

matsko committed:
fix(animations): always change to desired animation state even if no transition fires (#17025)

Fixes #16947

sjtrimble committed:
fix(aio): topnav toolbar position styles

- Fixed topnav on all mobile
- Fixed topnav on all docs pages
- Absolute topnav on all marketing pages
- Cleanup and code consolidation for all top-menu styles
- Add styling to topnav links on focus

sjtrimble committed:
fix(aio): swap feature headers for divs to remove anchors and fix feature image resizing (#17156)

sjtrimble committed:
fix(aio): fix home learn more button alignment (#17154)

chuckjaz committed:
feat(compiler-cli): produce template diagnostics error messages (#17125)

Refactoring the compiler to use transformers moves the code generation
after type-checking which suppresses the errors TypeScript would
generate in the user code.

`TypeChecker` currently produces the same factory code that was
generated prior the switch to transfomers, getting back the same
diagnostics as before. The refactoring will allow the code to
diverge from the factory code and allow better diagnostic error
messages than was previously possible by type-checking the factories.

matsko committed:
perf(animations): do not place enterId values on elements for querying purposes (#17150)

vicb committed:
docs: update the release schedule (#17162)

matsko committed:
fix(animations): do not retain deleted nodes during an non-removal animation (#17153)

Closes #17086

gkalpak committed:
fix(aio): fix scrolling to top (#17102)

Previously, the `#top-of-page` element (used when scrolling to top) was placed
inside the content section (which at the time had zero top margin and padding).
Furthermore, there was a top offset applied when scrolling that took the static
top bar's height into account. Since now the top bar is not static any more and
the content section has a non-zero top padding, scrolling to top does not work
as expected.

This commit fixes this by:
- Moving the `#top-of-page` element to the top of the `aio-shell`.
- Stop accounting for the top bar's top.

Fixes #17006

vicb committed:
release: cut the 4.2.0-rc.2 release

vicb committed:
docs: add changelog for 4.2.0-rc.2

alxhub committed:
feat: import of @angular/service-worker

Imports the existing repository (without history) from
https://github.com/angular/mobile-toolkit. Sets up the
package to build using build.sh and to run unit tests
as part of test.sh. An integration test is also added,
wrapping the SW e2e test so that `npm run test` works.

alexeagle committed:
build: Introduce Bazel build rules

So far this just compiles the core and common packages.

matsko committed:
perf(animations): only apply `:leave` flags if animations are set to run

matsko committed:
ci: update pullapprove to reflect animation ownership

matsko committed:
fix(animations): ensure web-animations understands a numeric CSS perspective value

Closes #14007

alexeagle committed:
ci: switch CircleCI to 2.0

    This lets us use a docker container to provide things like the Bazel build tool

rafaelss95 committed:
docs: fix typo in 'Routes' docs

Foxandxss committed:
build(aio): ignore .ngsummary.ts files in examples

alexeagle committed:
feat(compiler-cli): introduce synchronous codegen API

sjtrimble committed:
fix(aio): fix search box overlap for small devices (#17075)

- Adjust search box height
- Adjust search box standard width and width for smaller devices
- Fix search jump outside of specified max width

IgorMinar committed:
build(aio): upgrade ngo and purify (#17210)

petebacondarwin committed:
fix(aio): ensure all views can indicate the active node (#17194)

When more than one node matches a url, the last
node defined in the navigation.json file won. This
meant that, for instance, items in both the
TopBarNarrow and the Footer views would not
indicate that they were active.

Now, each url is associated with a map of current
nodes keyed off their view.

Closes #17022

petebacondarwin committed:
refactor(aio): use the SelectMenuComponent for all select menus

The API filters and the docs version switcher now use
the SelectMenuComponent.

Fixes #16367 and #17055

Foxandxss committed:
build(aio): update features and presskit marketing pages

Foxandxss committed:
build(aio): migrate contribute page to html

Foxandxss committed:
docs(aio): remove tutorial from docs landing page

gkalpak committed:
fix(aio): fix and clean up topbar styling

Restore the changes introduced in #17075, which wre accidentally overwritten
while rebasing #17155. Also, simplify the topbar positioning rules.

gkalpak committed:
fix(aio): make `ScrollSpy` respond quicker to `scroll` events

Fixes ##17220

sjtrimble committed:
fix(aio): fix button alignment and jump

- Fix alignment on get started button on home page
- Remove jump by removing arrow from learn more button on home page

gkalpak committed:
fix(aio): fix scrolling to an element

Previously, the top-bar's height wasn't taken into account when scrolling an
element into view. As a result, the element would be hidden behind the top-bar.
Taking the top-bar height into account was not necessary before #17155, because
the top-bar was not fixed (i.e. it scrolled away).

This commit fixes the scrolling behavior by accounting for the top-bar's height
when scrolling an element into view.

(This partially reverts #17102.)

Fixes #17219
Fixes #17226

matsko committed:
perf(animations): do not create a closure each time a node is removed

mary-poppins commented:
The angular.io preview for 98a84a959ef9afbe0045d56c158aef894f523486 is available [here][1].

[1]: https://pr17276-98a84a959ef9afbe0045d56c158aef894f523486.ngbuilds.io/

gkalpak committed:
fix(aio): adjust SideNav top to avoid extra whitespace

Partially addresses #17096.

alxhub committed:
fix(http): move destructuring inside {Request,Response}Options ctor

Previously the RequestOptions/ResponseOptions classes had constructors
with a destructured argument hash (represented by the
{Request,Response}OptionsArgs type). This type consists entirely of
optional members.

This produces a .d.ts file which includes the constructor declaration:

constructor({param, otherParam}?: OptionsArgs);

However, this declaration doesn't type-check properly. TypeScript
determines the actual type of the hash parameter to be OptionsArgs | undefined,
which it then concludes does not have a `param` or `otherParam` member.

This is a bug in TypeScript ( https://github.com/microsoft/typescript/issues/10078 ).
As a workaround, destructuring is moved inside the method, where it does not produce
broken artifacts in the .d.ts.

Fixes #16663.

wardbell committed:
fix(aio): re-color code pretty print

closes #17246

petebacondarwin committed:
fix(aio): remove `...` separator from search results

An ellipsis was used to separate the most relevant search
results from the alphabetic list. The separator was confusing
because it was not clear what it represented.

This has been removed and the most relevant results are now
indicated by styling with a more bold font and a bit of whitespace
between them and the rest of the results.

To keep things consistent, if there are fewer than 5 results all the
results are now displayed as priorityPages.

Closes #17233

petebacondarwin committed:
docs(aio): use h1 rather than divs and hide heading anchors

In the marketing pages we do not want to show heading anchors on hover.
Previously, this was achieved by using div rather than heading elements.
Now we can use semantically accurate headings while hiding the anchor.

Closes #17244
Closes #17264

gkalpak committed:
fix(aio): expand the main content width when there is no ToC

Previously, the main content would always leave a 18% margin on the right to be
occupied by the ToC (even if there was no ToC).
This commit lets the main content expand to the right to occupy all the
available space when there is no ToC.

Fixes #17205
Fixes #17270

petebacondarwin committed:
feat(aio): do not display anchor if heading has `no-anchor` class

sjtrimble committed:
fix(aio): update contrast on mediumgray color

petebacondarwin committed:
feat(aio): add `aio-select` component

Provide the functionality for select menus in a single reusable component.

wardbell committed:
docs: security guide images
closes #17263

wardbell committed:
docs(aio): add “nougat” to compat chart

closes #17262

wardbell committed:
docs: "TypeScript to JavaScript" code tab renamed “ES6 + Decorators”

closes #17260

petebacondarwin committed:
build(aio): ignore example files that are gitignored

Rather than hard coding excludes into the dgeni config,
use the fact that we are already ignoring the boilerplate
and generated files via the .gitignore file.

matsko committed:
fix(animations): evaluate substitutions on option param values

mary-poppins commented:
The angular.io preview for 2d0c1127aef2208cdbc7b0a5f8ccd135cb9491fb is available [here][1].

[1]: https://pr17276-2d0c1127aef2208cdbc7b0a5f8ccd135cb9491fb.ngbuilds.io/

mhevery committed:
feat(core): update zone.js to 0.8.12

mary-poppins commented:
The angular.io preview for c74e922cd41a8764f208d4f385cc1670c893a39c is available [here][1].

[1]: https://pr17276-c74e922cd41a8764f208d4f385cc1670c893a39c.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 6d4e3a5a085e463a08dbd54836e55283cf220080 is available [here][1].

[1]: https://pr17276-6d4e3a5a085e463a08dbd54836e55283cf220080.ngbuilds.io/

mary-poppins commented:
The angular.io preview for afd0f2f757a1047f4d555a11992ced0b69bbda75 is available [here][1].

[1]: https://pr17276-afd0f2f757a1047f4d555a11992ced0b69bbda75.ngbuilds.io/

kara committed:
fix(forms): fix min and max validator behavior on non-numbers

IgorMinar committed:
build(aio): turn on webpack 3.0 scope-hoisting

IgorMinar committed:
build(aio): turn off vendor chunking

This was done to prevent webpack from generating cross chunk registry calls which prevent DCE.

IgorMinar committed:
build(aio): upgrade to purify 0.0.24 that supports webpack 3.0

IgorMinar committed:
build(aio): upgrade @angular/service-worker to beta.15

IgorMinar committed:
build(aio): upgrade to an unreleased version of cli with webpack 3.0-rc.0

petebacondarwin committed:
docs(aio): do not ignore i18n example js file

petebacondarwin committed:
docs(aio): move example specific styles into router example app.css file

petebacondarwin committed:
docs(aio): remove example specific styles from the shared boilerplate styles.css

gkalpak committed:
fix(aio): smoother transition from page with SideNav to homepage (#17312)

When navigating from a page with open SideNav to a page without closed SideNav,
the main content area animates from a non-zero left margin to zero left margin.
Additionally, the top-bar on the homepage is transparent, which allows the white
background behind the main content to be seen while the left margin is animated
to zero, making it appear as if something (e.g. the SideNav covers the top-bar).

This commit works around this issue, by not making the top-bar transparent
immediately when navigating to the homepage, but animating it from its blue
color to transparent with a delay.

Fixes #17248

gkalpak committed:
fix(aio): make the ToC left hand-side bar shorter (#17280)

Make the ToC left hand-side bar shorter, so that it starts at the initial bullet
and ends at the final bullet.

Fixes #17221

petebacondarwin committed:
docs(aio): remove cli-quickstart styles.css from git

olegdunkan committed:
refactor(compiler): remove duplicate code

gkalpak committed:
ci: update Chromium to 464641 (Chrome v59)

gkalpak committed:
fix(aio): restrain scrolling inside ToC (when cursor over ToC)

Previously, when scrolling the ToC and reaching the top/bottom, further
mousewheel events would result in scrolling the window (and thus the main
content). This is standard browser behavior. In the case of the ToC though, the
`ScrollSpy` would detect scrolling in the main content and scroll the active ToC
to entry into view, thus resetting the scroll position of the ToC.

Reproduction:
1. Open  `~/guide/template-syntax`.
2. Start scrolling through the long ToC.
3. Try to go to the bottom of the ToC.
4. Once you reach the bottom, the main content starts scrolling down.
5. The first section ("HTML in templates") becomes "active", so the ToC is
   scrolled back up to make its corresponding entry visible.
6. Go back to step 2.

This commit improves the UX, by not allowing the main content to scroll when the
cursor is ovr the ToC and the user has scrolled all the way to the top/bottom of
it.

gkalpak committed:
fix(aio): animate hamburger in/out

Fixes #17215

petebacondarwin committed:
docs(aio): create styles.css snippet file for toh-pt5 example

Foxandxss committed:
build(aio): rename features header

mary-poppins commented:
The angular.io preview for 31f0fe0adee0646d95043686cc16b2bb9f7a6b10 is available [here][1].

[1]: https://pr17276-31f0fe0adee0646d95043686cc16b2bb9f7a6b10.ngbuilds.io/

sjtrimble committed:
fix(aio): api list column overflow fix (#17300)

wardbell committed:
fix(aio): scroll to top immediately when doc changes (#17310)

sjtrimble committed:
fix(aio): alert, subsection and callout width fix (#17303)

sjtrimble committed:
fix(aio): remove glossary l-sub-section class (#17305)

sjtrimble committed:
fix(aio): contributor img offset and hover fix (#17338)

sjtrimble committed:
fix(aio): copy code button overlap fix on mobile (#17302)

sjtrimble committed:
fix(aio): style and fix hover color for features cta button (#17342)

petebacondarwin committed:
docs(aio): remove example specific `.gitignore` files

All the files that should be ignored are defined in the
`aio/content/examples/.gitigore` folder.

alxhub committed:
release: cut the 4.2.0 release

alxhub committed:
docs: add changelog for 4.2.0

IgorMinar committed:
feat(aio): update the home page and docs landing page content (#17325)

Closes #17268
Closes #17230

wardbell committed:
fix(aio): sidebar folder state after select item

Closes #17245 and #17253

When the user selects a doc item in the side nav:
1) expand folder(s) leading to the selected doc item
2) on a wide display, keep other already expanded folders open
3) on narrow (mobile) display, collapse other expanded folders

Used to do (3) when wide. Issue #17245 asks for (2).

That logic was bypassed for selected node when we allowed headers to have content
because that unintentionally expanded the header’s folder when selected.
Because the selected node is no longer a header with content, removing this exclusion
also means that folders are expanded/collapsed with above logic even for API pages.

trotyl committed:
fix(aio): fix trackBy demo in template-syntax article

gkalpak committed:
fix(aio): temporarily remove the focus style of top-bar nav items

Temporarily addresses #17216 until we upgrade to an `@angular/material` version
that includes angular/material2@3bc82f6dc.

IgorMinar committed:
docs(aio): update events page

IgorMinar committed:
docs(aio): fix image width on the homepage

frederikprijck committed:
docs(aio): update typescript for `examples/webpack` to same as cli

Foxandxss committed:
build(aio): update to 4.2 (#17369)

dherges committed:
docs(packaging): update homepage in @angular/tsc-wrapped (#17352)

Motivation: `yarn outdated`, for exmaple, shows the homepage URL on the command line. If copy-pasting or clicking on the URL, it's nice to see the repo's page instead of a 404.

tbosch committed:
refactor(compiler): don’t write summaries for jit by default

The default is false externally but true internally at Google.

gkalpak committed:
fix(aio): temporarily remove link to source from the API pages (#17371)

We need to come up with a better design (possibly involving an icon button) to
link to the source code (for viewing and/or editing).

Fixes #17254

Foxandxss committed:
docs(aio): update examples to 4.2 (#17367)

Foxandxss committed:
build(aio): update typescript to 2.3.2 (#17382)

matsko committed:
fix(animations): compute removal node height correctly

kapunahelewong committed:
docs(aio): reorganise the sidenav menu (#16934)

Reorganizes the items in the sidenav menu and consolidates the quickstart
and cli-quickstart guides into one.

matsko committed:
fix(animations): do not treat a `0` animation state as `void`

IgorMinar committed:
build(aio): remove duplicate webpack version

alxhub committed:
release: cut the 4.2.1 release

alxhub committed:
docs: add changelog for 4.2.1

IgorMinar committed:
build(aio): upgrade to @angular/*@4.2.1

IgorMinar committed:
build(aio): upgrade purify to deal with messed up comment in animations code

IgorMinar committed:
fix(animations): remove duplicate license header

Foxandxss committed:
docs(aio): restore animation original value (#17385)

IgorMinar committed:
build(bazel): fix the typescript path in the BUILD file (#17316)

petebacondarwin committed:
fix(aio): remove gap between sidenav menus

Fixes #17394

gkalpak committed:
build(aio): upgrade Karma-related dependencies

gkalpak committed:
build(aio): upgrade Jasmine-related dependencies

According to https://github.com/jasmine/jasmine/issues/1327#issuecomment-306924549,
this might help with help with frequent DISCONNECTED errors (especially on Travis).

petebacondarwin committed:
fix(aio): remove outline from search input on focus

Closes #17396

petebacondarwin committed:
fix(aio): tidy up layout of api filter page

* Remove the "info-banner" styling from the filters.
* Fix alignment of the search box on a narrow screen (closes #17395)
* Remove unnecessary whitespace before section headers

petebacondarwin committed:
fix(aio): ensure that API filter page can display 3 columns in wide view

Fixes #17251

gkalpak committed:
test(aio): fix Chrome getting disconnected during unit tests

Chrome (v58+) often gets disconnected during unit tests (causing them to fail).
This has been happening locally (on Windows) and on Travis. The exact reasons
are not known, but it seems that some of the `AppComponent` tests are to blame.

Based on the discussion in https://github.com/jasmine/jasmine/issues/1327 (and
plenty of trial-and-error), using Jasmine's `done()` callback before each of
test (even calling it synchronously) fixes the issue.

alxhub committed:
fix: argument destructuring sometimes breaks strictNullChecks

Destructuring of the form:

function foo({a, b}: {a?, b?} = {})

breaks strictNullChecks, due to the TypeScript bug https://github.com/microsoft/typescript/issues/10078.
This change eliminates usage of destructuring in function argument lists in cases where it would leak
into the public API .d.ts.

matsko committed:
fix(animations): properly collect :enter nodes in a partially updated collection

This PR fixes an issue where `query(':enter')` will only collect elements up until it an element that is found that isn't apart of the `:enter` query.

Closes #17440

wardbell committed:
docs: clarify when non-null-assertion-operator is needed in template-syntax

juleskremer committed:
docs(aio): rename Upgrade docs to cheatsheet

juleskremer committed:
docs(aio): update about page

petebacondarwin committed:
fix(aio): use locally hosted lunr library

The library is downloaded from npm but then
copied into the assets folder (and ignored by git)
as part of the postinstall step.

tbosch committed:
refactor(compiler): don’t always compile `.ngfactory.ts` files by default

This puts the behavior introduced in https://github.com/angular/angular/commit/573b8611bc4ab85a7436ee110526678a0ecf6c76 behind the new flag
`alwaysCompileGeneratedCode` to not break users that might have relied
on this behavior.

petebacondarwin committed:
fix(aio): show search results when search box gets focus

Due to a previous commit, the search was only triggered
if the query changed, and not when the search box regained
focus.

IgorMinar committed:
build(aio): update Google Analytics property id for the prod environment

This is the same property ID as the one used by the current angular.io.

alxhub committed:
release: cut the 4.2.2 release

alxhub committed:
docs: add changelog for 4.2.2

petebacondarwin committed:
build(aio): add staging environment

You can now specify what environment you are building
by add it to the `yarn build` command. For example:

```
yarn build -- --env=stage
```

Moreover the `deploy-to-firebase.sh` script will automatically apply the
appropriate environment.

alexeagle committed:
ci: use npm_install for bazel

using yarn_install polluted our node_modules cache because it disregards the npm_shrinkwrap.json

IgorMinar committed:
build(aio): make deploy-to-firebase.sh executable

gkalpak committed:
ci(aio): fail the build if the PWA score is too low

Previously, there was an issue with testing the PWA score on staging and failing
the build was temporarily disabled. It works now, so we need to enable failing
the build is the score drops below some threshold.

gkalpak committed:
build: remove redundant line

The same value is set a few lines below.

gkalpak committed:
build(aio): remove dependency on build artifacts from parent folder

gkalpak committed:
fix(aio): fix buttons in "Home" and "Features"

Using `<a>` inside a `<button>` is not syntactically valid HTML and breaks on
some browsers (e.g. Firefox). Furthermore, clicking the button doesn't do
anything unless you click on the link (e.g. clicking on the padding around the
link does nothing), which is inconvenient and confusing.

Fixes #17448

dherges committed:
docs(aio): http guide shows how to import toPromise operator from rxjs

Solves #17454

petebacondarwin committed:
fix(aio): make search results better

* update to latest version of lunr search
* add trailing wildcard to search terms to increase matches
* fix unwanted error when escape was pressed

Closes #17417

gkalpak committed:
fix(aio): fix scrolling to elements near the bottom of the page

Previously, we always assumed that elements would be scrolled to the top of the
page, when calling `element.scrollIntoView()`. This is not true for elements
that cannot be scrolled to the top, e.g. when the viewport height is larger than
the height of the content after the element (common for small sections near the
end of the page).
In such cases, we would unnecessarily scroll up to account for the static
toolbar, which was unnecessary (since the element was not behind the toolbar
anyway) and caused ScrollSpy to fail to identify the scrolled-to section as
active.

This commit fixes it by ensuring that we do not scroll more than necessary in
order to align the top of the element with the bottom of the toolbar.

Fixes #17452

gkalpak committed:
fix(aio): make the footer links clickable on all browsers

The footer background (implemented via `footer:after`) had a higher `z-index`
than other footer elements and was obscuring the footer links on certain
browsers (Firefox, Edge, IE), which made them unclickable.
This commit lowers the index of `footer:after`, so that links are clickable on
these browsers.

Fixes #17460

petebacondarwin committed:
build(aio): increase docs integration test timeouts

The API docs tests have very variable run times, depending
upon the build environment.
This change doubles their test timeout values to prevent
false-negative failures.

Foxandxss committed:
build(aio): extra redirect rule

gkalpak committed:
build(aio): upgrade lighthouse to v2.1

juleskremer committed:
docs(aio): update resource listing

gkalpak committed:
fix(aio): specify large image for PWA splash-screen

alxhub committed:
ci: disable platform-server integration test as it is currently broken

vikerman committed:
test(platform-server): fix and re-enable integration tests

Foxandxss committed:
ci: update github templates (#17466)

gkalpak committed:
fix(aio): always cover the whole footer with its background

Fixes #17465

mary-poppins commented:
The angular.io preview for a719cc25fbb59a1346905d7891d6e95e9ed98e6b is available [here][1].

[1]: https://pr17276-a719cc25fbb59a1346905d7891d6e95e9ed98e6b.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 534ddb51d1b17fc00139383dfeda8322961efd02 is available [here][1].

[1]: https://pr17276-534ddb51d1b17fc00139383dfeda8322961efd02.ngbuilds.io/

petebacondarwin committed:
test(aio): add async beforeEach to prevent Chrome disconnects

Related to 3d5f520ff08f00f89f323f1ec3bcd1e0c3f38596 from #17405

gkalpak committed:
build: fix link to DEVELOPER.md in `check-environment.js`

gkalpak committed:
build(aio): do not fail if `check-env` for the main angular project fails

Fixes #17434

kapunahelewong committed:
docs(aio): add short section on built-in validators, copy edits

gkalpak committed:
refactor(aio): remove unused `Logger` dependencies

gkalpak committed:
fix(aio): do not log messages in production

In dev mode, all messages passed to `Logger` will be logged.
In production mode, only warnings and errors will be logged.

Fixes #17453

kapunahelewong committed:
docs(aio): incorporate Ward's comments

jesselpalmer committed:
feat(aio): add iphone pwa features

devversion committed:
docs(aio): re-add biography entry for devversion

With SHA 2c3e948e61a22da9e0612f882edfb830e8470b6c the biography of Paul Gschwendtner has been accidentally removed.

This re-adds the biography entry (picture still present) as requested on Slack.

pkozlowski-opensource committed:
refactor(core): remove toString() method from DefaultKeyValueDiffer

toString() from DefaultKeyValueDiffer is only used in tests and should not
be part of the production code. toString() methods from differs add
~ 0.3KB (min+gzip) to the production bundle size.

gkalpak committed:
fix(aio): add missing redirect rule for `/styleguide`

Fixes #17542

IgorMinar committed:
fix(router): update the version placeholder so that it gets replaced during the build

Fixes #17403

kara committed:
fix(forms): temp roll back breaking change with min/max directives

With 4.2, we introduced the min and max validator directives. This was actually a breaking change because
their selectors could include custom value accessors using the min/max properties for their own purposes.

For now, we are rolling back the change by removing the exports.

Closes #17491.

gkalpak committed:
fix(aio): correctly redirect `/docs/ts/latest` and `/styleguide`

Previously, we had redirect rules for Firebase for `/docs/ts/latest` and
`/styleguide`, but once the ServiceWorker was activated, it would take over
routing and rewrite these requests to `/index.html`.
This commit fixes it by excluding them from ServiceWorker routing.

Fixes #17542

gkalpak committed:
fix(aio): remove unused icon reference and file

These icons are not there (and never were afaict).

Fixes #17561

gkalpak committed:
fix(aio): improve no-javascript screen

mary-poppins commented:
The angular.io preview for 7b346699af80f979de95595c6c7f7ac9357815b1 is available [here][1].

[1]: https://pr17276-7b346699af80f979de95595c6c7f7ac9357815b1.ngbuilds.io/

hansl committed:
docs: add changelog for 4.2.3

mary-poppins commented:
The angular.io preview for 2fc73b61adf382dc003d685f229a640ae9d8f163 is available [here][1].

[1]: https://pr17276-2fc73b61adf382dc003d685f229a640ae9d8f163.ngbuilds.io/

alxhub committed:
build(aio): update @angular/service-worker to 1.0.0-beta.16

This version includes a fix for potential cache corruption and invalid redirect behavior in rare cases.

gkalpak committed:
fix(aio): leave results panel open when opening search result on new page

Fixes #17580

gkalpak committed:
build(aio): upgrade jasmine to v2.6.4

This version fixes the DISCONNECTED errors (described in #17543) and removes the
need to the workaround (8af203c).
The relevant jasmine commit is jasmine/jasmine@c60d66994.

gkalpak committed:
fix(aio): switch from `innerText` to `textContent` to support older browsers

`innerText` is not supported in Firefox prior to v45. In most cases (at least
the ones we are interested in), `innerText` and `textContent` work equally well,
but `textContent` is more performant (as it doesn't require a reflow).

From [MDN][1] on the differences of `innerText` vs `textContent`:

> - [...]
> - `innerText` is aware of style and will not return the text of hidden
>   elements, whereas `textContent` will.
> - As `innerText` is aware of CSS styling, it will trigger a reflow, whereas
>   `textContent` will not.
> - [...]

[1]: https://developer.mozilla.org/en-US/docs/Web/API/Node/textContent#Differences_from_innerText

Fixes #17585

chuckjaz committed:
fix(language-service): infer `any` `ngForOf` of type `any`

Fixes: #17611

IgorMinar committed:
docs(RELEASE_SCHEDULE): update the release schedule w/ recent regression patch releases

chuckjaz committed:
fix(language-service): rollup `tslib` into the language service package

Fixes: #17614

hansl committed:
docs: remove aio from changelog

IgorMinar committed:
docs(TRIAGE_AND_LABELS): update labels to reflect the current state

IgorMinar committed:
docs(RELEASE_SCHEDULE): fix version numbers for August/September releases

hansl committed:
revert: fix(forms): temp roll back breaking change with min/max directives

This reverts commit 232bd9395dbae3a22467aa669126d50cbaa7ce56.

hansl committed:
fix(forms): roll back breaking change with min/max directives

With 4.2, we introduced the min and max validator directives. This was actually a breaking change because their selectors could include custom value accessors using the min/max properties for their own purposes.

For now, we are rolling back the change by removing the exports. At the least, we should wait to add them until a major version. In the meantime, we will have further discussion about what the best solution is going forward for all validator directives.

Closes #17491.

----

PR #17551 tried to roll this back, but did not remove the dead code. This failed internal tests that were checking that all declared directives were used.
This PR rolls back the original PR and commit the same as #17551 while also removing the dead code.

IgorMinar commented on integration/service-worker/src/app/app.component.html:
```diff
@@ -0,0 +1,2 @@
+<h1>Service Worker Testing Harness</h1>
+<controller></controller>
```
can you please prefix the component selector with `test-` or similar?

alxhub replied:
Done.

chuckjaz committed:
fix(tsc-wrapped): skip collecting metadata for default functions

Fixes: #17518

IgorMinar commented on npm-shrinkwrap.json:
```diff
@@ -1,549 +1,655 @@
 {
   "name": "angular-srcs",
-  "version": "4.2.0-beta.1",
+  "version": "4.2.0-rc.2",
```
can you undo all the changes unnecessary changes or redo the shrinkwrap file with npm 3.10.7 as required by env.sh?

otherwise we have all this noise in the shrinkwrap file. unless ci is updated to use newer version of npm we shouldn't use newer one locally.

IgorMinar commented on package.json:
```diff
@@ -68,6 +68,7 @@
     "jasmine": "^2.4.1",
     "jasmine-core": "^2.4.1",
     "jpm": "1.0.0",
+    "jshashes": "^1.0.6",
```
what is this dependency for?

3rd party dependencies are tricky. Especially if they make it into the browser.

Also if we use it just for SHA1 then we already have an implementation of that that Vic wrote for i18n, it would be simpler to use that instead. Let's discuss.

IgorMinar commented on packages/service-worker/rollup.config.js:
```diff
@@ -0,0 +1,27 @@
+ */
+
+import resolve from 'rollup-plugin-node-resolve';
+import commonJs from 'rollup-plugin-commonjs';
```
what do we need this plugin for? other of our packages don't need it.

alxhub replied:
`jshashes`

chuckjaz committed:
fix(compiler-cli): find lazy routes in nested module import arrays

Fixes: #17531

mary-poppins commented:
The angular.io preview for d4324d2f6938f2d6d70e048e22a719e36c869afe is available [here][1].

[1]: https://pr17276-d4324d2f6938f2d6d70e048e22a719e36c869afe.ngbuilds.io/

IgorMinar commented on packages/service-worker/package.json:
```diff
@@ -0,0 +1,21 @@
+{
+  "name": "@angular/service-worker",
+  "version": "0.0.0-PLACEHOLDER",
+  "description": "Angular - service worker implementation",
```
"implementation" -> "support and tooling"?

alxhub replied:
Done

alxhub committed:
feat: add builds to @angular/service-worker

IgorMinar commented on tools/public_api_guard/service-worker/service-worker.d.ts:
```diff
@@ -0,0 +1,36 @@
+/** @experimental */
+export declare class NgPushRegistration {
+    readonly url: string;
+    constructor(ps: any);
```
shouldn't this be an interface or an abstract class? we don't want people to instantiate it on their own.

alxhub replied:
I removed the constructor from the public API via `@internal`.

IgorMinar commented on tools/public_api_guard/service-worker/service-worker.d.ts:
```diff
@@ -0,0 +1,36 @@
+/** @experimental */
+export interface UpdateEvent {
+    type: 'pending' | 'activation';
+    version?: string;
```
is it true that this is optional?

IgorMinar commented on tools/public_api_guard/service-worker/service-worker.d.ts:
```diff
@@ -0,0 +1,36 @@
+}
+
+/** @experimental */
+export interface PushOptions {
```
could we inline this simple type?

IgorMinar commented on tools/public_api_guard/service-worker/service-worker.d.ts:
```diff
@@ -0,0 +1,36 @@
+    readonly url: string;
+    constructor(ps: any);
+    auth(): string;
+    key(method?: string): string;
```
do we need to expose these `auth` and `key` strings? the naming of them is confusing, but a more important question is if we should expose this at all.

alxhub replied:
Those two are the values that need to be sent to the server to support sending pushes to the client.

IgorMinar commented on tools/public_api_guard/service-worker/service-worker.d.ts:
```diff
@@ -0,0 +1,36 @@
+    constructor(ps: any);
+    auth(): string;
+    key(method?: string): string;
+    toJSON(): Object;
```
what's the purpose of this?

alxhub replied:
To get subscription credentials that allow sending push notifications to the client.

IgorMinar commented on tools/public_api_guard/service-worker/service-worker.d.ts:
```diff
@@ -0,0 +1,36 @@
+    auth(): string;
+    key(method?: string): string;
+    toJSON(): Object;
+    unsubscribe(): Observable<boolean>;
```
this looks odd.

why do we have two streams and a special unsubscribe API?

why isn't the api just:

```
export declare class NgServiceWorker {
  pushNotifications(options: {applicationServerKey?: string}): Observable<ConnectionInfo|Notification>;
  updateNotifications(): Observable<UpdateEvent>;
  checkForUpdate(): void;
  activateUpdate(): void;
}
```

This would simplify the registration and  funnel all appropriate events through the two notification streams.

alxhub replied:
Observables are values. Since they're singletons and not parameterized in their creation, they should be fields. I want to rethink this whole API at some point.

IgorMinar commented on tools/public_api_guard/service-worker/service-worker.d.ts:
```diff
@@ -0,0 +1,36 @@
+    constructor(zone: NgZone);
+    activateUpdate(version: string): Observable<boolean>;
+    checkForUpdate(): Observable<boolean>;
+    log(): Observable<string>;
```
is the logging functionality meant to be a public api?

what are the use cases for this?

IgorMinar commented on tools/public_api_guard/service-worker/sdk/plugins.d.ts:
```diff
@@ -0,0 +1,50 @@
+export declare function StaticContentCache(options?: StaticContentCacheOptions): PluginFactory<StaticContentCacheImpl>;
+
+/** @experimental */
+export interface StaticContentCacheOptions {
```
inline?

alxhub replied:
Done

IgorMinar commented on tools/public_api_guard/service-worker/sdk/plugins.d.ts:
```diff
@@ -0,0 +1,50 @@
+}
+
+/** @experimental */
+export declare function Push(): PluginFactory<PushImpl>;
```
`PushNotifications` ?

alxhub replied:
Future refactor :)

IgorMinar commented on tools/public_api_guard/service-worker/sdk/plugins.d.ts:
```diff
@@ -0,0 +1,50 @@
+/** @experimental */
+export declare function Dynamic(strategies: DynamicStrategy[]): PluginFactory<DynamicImpl>;
```
rename to `DynamicContentCache` to be consistent with `StaticContentCache`?

alxhub replied:
Will do in a future refactor (plugins are not public API now either).

IgorMinar commented on tools/public_api_guard/service-worker/sdk/plugins.d.ts:
```diff
@@ -0,0 +1,50 @@
+}
+
+/** @experimental */
+export declare class FreshnessStrategy implements DynamicStrategy {
```
`DynamicStrategy` is not exported

alxhub replied:
Good catch. Fixed.

IgorMinar commented on tools/public_api_guard/service-worker/sdk.d.ts:
```diff
@@ -0,0 +1,284 @@
+}
+
+/** @experimental */
+export declare class VersionWorkerImpl implements VersionWorker {
```
do we need to export the implementation in addition to the interface?

alxhub replied:
Yeah, the two are not currently in sync. Will fix in a future refactoring.

IgorMinar commented on tools/public_api_guard/service-worker/sdk.d.ts:
```diff
@@ -0,0 +1,284 @@
+export declare function parseManifest(data: string): Manifest;
+
+/** @experimental */
+export interface Plugin<T extends Plugin<T>> {
```
this sounds like a too generic name. rename to NgSwPlugin to follow the existing conventions?

alxhub replied:
Will do when I refactor later.

IgorMinar commented on tools/public_api_guard/service-worker/sdk.d.ts:
```diff
@@ -0,0 +1,284 @@
+/** @experimental */
+export interface Manifest {
+    _hash: string;
+    _json: string;
```
should the _ props be public? if so, why have the prefix?

alxhub replied:
So they don't clash with any user-defined properties on the manifest.

IgorMinar commented on tools/public_api_guard/service-worker/sdk.d.ts:
```diff
@@ -0,0 +1,284 @@
+/** @experimental */
+export declare class BrowserClock implements Clock {
+    dateNow(): number;
+    setTimeout(fn: Function, delay: number): number;
```
it's a bit odd that Clock api has a setTimeout method. That's a job for a scheduler not a clock. Is do we need this to be a public api?

alxhub replied:
It's not public API anymore. I'll refactor it later, I have plans to clean up and reduce the number of internal facades the SW uses for testing.

IgorMinar commented on tools/public_api_guard/service-worker/sdk.d.ts:
```diff
@@ -0,0 +1,284 @@
+/** @experimental */
+export interface BootstrapOptions {
```
can we inline this type? it's easier to write the api docs when these kind of options are on the same page as the method/fn signature.

additionally, we don't ever want anyone to implement these kinds of interfaces because then it's harder to change them without a breaking change.

alxhub replied:
Done. SDK was removed from public API anyway.

alxhub replied:
Do we have a pattern for documenting the members of such inlined interfaces?

IgorMinar commented on packages/service-worker/zprebuilt/src/prebuilt.ts:
```diff
@@ -0,0 +1,20 @@
+import {bootstrapServiceWorker} from '@angular/service-worker/sdk';
+import {Dynamic, ExternalContentCache, FreshnessStrategy, PerformanceStrategy, Push, RouteRedirection, StaticContentCache} from '@angular/service-worker/sdk/plugins';
+
+export function main() {
```
remove main? it's a dart artifact.

alxhub replied:
Done.

alxhub commented on integration/service-worker/package.json:
```diff
@@ -0,0 +1,47 @@
+{
+  "name": "service-worker",
```
We discussed this in person - currently the integration tests remove any lockfile. This should be fixed upstream

alxhub commented on integration/service-worker/e2e/server/push.ts:
```diff
@@ -0,0 +1,24 @@
+export function sendPush(subscription: any, payload?: Object): Promise<any> {
+  return push.sendNotification(subscription, JSON.stringify(payload), {
+    vapidDetails: {
+      subject: 'mailto:test@angular.io',
```
Just a fake value - and the script itself can probably be deleted, as we no longer use the real Push API to test.

alxhub commented on integration/service-worker/e2e/server/push.ts:
```diff
@@ -0,0 +1,24 @@
+  return push.sendNotification(subscription, JSON.stringify(payload), {
+    vapidDetails: {
+      subject: 'mailto:test@angular.io',
+      publicKey: 'BLRl_fG1TCTc1D2JwzOpdZjaRcJucXtG8TAd5g9vuYjl6KUUDxgoRjQPCgjZfY-_Rusd_qtjNvanHXeFvOFlxH4',
```
Yes. Totally safe to commit to a Github repo - it identifies the application. Only someone with the private key can push to subscribers to the app. Since the app is a test, nobody cares.

alxhub commented on integration/service-worker/protractor.config.js:
```diff
@@ -0,0 +1,42 @@
+    /**
+     * ng2 related configuration
+     *
+     * useAllAngular2AppRoots: tells Protractor to wait for any angular2 apps on the page instead of just the one matching
```
Sure. This is a CLI template file, though.

alxhub replied:
Done.

alxhub committed:
feat: use flux standard actions for worker <-> client communication

mary-poppins commented:
You can preview c364455 at https://pr17276-c364455.ngbuilds.io/.

mary-poppins commented:
You can preview b267b24 at https://pr17276-b267b24.ngbuilds.io/.

mary-poppins commented:
You can preview 1b30a81 at https://pr17276-1b30a81.ngbuilds.io/.

alxhub commented on integration/service-worker/e2e/server/server.ts:
```diff
@@ -0,0 +1,70 @@
+import express = require('express');
```
Done

alxhub commented on packages/service-worker/sdk/test/worker_spec.ts:
```diff
@@ -0,0 +1,308 @@
+import 'reflect-metadata';
```
Probably not. Removed.

mary-poppins commented:
You can preview 5c3f29a at https://pr17276-5c3f29a.ngbuilds.io/.

mary-poppins commented:
You can preview 712c269 at https://pr17276-712c269.ngbuilds.io/.

alxhub committed:
refactor: address comments in code review

IgorMinar commented:
@alxhub can you rebase this PR plz?

jo3d3v commented:
Is this PR planned for angular 5 or later? It would be nice to know, when mobile toolkit (or parts) is merged in the main repo. I plan to build up different PWAs on that module. 
While playing with mobile toolkit I identified different enhancements or extensions. (Posted in https://github.com/angular/mobile-toolkit/issues/138#issuecomment-300905267) I would like to support this ideas with appropriate PRs on that module.

somombo commented:
In light of https://github.com/angular/angular/pull/19274 is this PR still in the works?
